### PR TITLE
Issues 19395 19396 19397

### PR DESCRIPTION
**Pull request by nollymar**
New fields are included into the `identifier` table. These are: `owner`, `create_date` and `asset_subtype`

### CODE REVIEW

**Review by freddyucv — CHANGES_REQUESTED** [review]
any test?

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public boolean forceRun() {
+        return true;
+    }
+
+    @Override
+    public String getPostgresScript() {
+        return "alter table identifier add owner varchar(255);" +
+                "alter table identifier add create_date timestamp;" +
+                "alter table identifier add asset_subtype varchar(255);";
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:16]: asset_subtype needs an index on it

File: dotCMS/src/main/resources/mssql.sql [hunk 2]
```diff
@@ -1266,6 +1266,9 @@ create table identifier (
    syspublish_date datetime null,
    sysexpire_date datetime null,
    full_path_lc  as CASE WHEN parent_path = 'System folder' THEN '/' ELSE LOWER(CONCAT(parent_path, asset_name)) END,
+   owner NVARCHAR(255),
+   create_date datetime,
+   asset_subtype NVARCHAR(255),
```
Comment (wezell) [inline dotCMS/src/main/resources/mssql.sql:1271]: add index to `identifier.asset_subtype`

**Review by wezell — CHANGES_REQUESTED** [review]
Also, we need the startup task that populates the new fields.

File: dotCMS/src/main/java/com/dotmarketing/beans/transform/IdentifierTransformer.java [hunk 3]
```diff
@@ -42,6 +42,9 @@ private static Identifier transform(Map<String, Object> map)  {
         i.setParentPath((String) map.get("parent_path"));
         i.setSysPublishDate((Date) map.get("syspublish_date"));
         i.setSysExpireDate((Date) map.get("sysexpire_date"));
+        i.setOwner((String) map.get("owner"));
+        i.setCreateDate((Date) map.get("create_date"));
+        i.setAssetSubType((String) map.get("asset_subtype"));
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/beans/transform/IdentifierTransformer.java]: Done

**Comment by nollymar** [comment]
> any test?

Done

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java (hunk 1 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java]: Done

File: dotCMS/src/main/resources/mssql.sql (hunk 2 above)
Comment (nollymar) [reply dotCMS/src/main/resources/mssql.sql:1271]: Done

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    }
+
+
+    @Override
+    public String getPostgresScript() {
+        return "alter table identifier add owner varchar(255);" +
+                "alter table identifier add create_date timestamp;" +
+                "alter table identifier add asset_subtype varchar(255);" +
+                "create index idx_identifier_asset_subtype on identifier (asset_subtype);";
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:29]: Codacy found an issue: [The String literal "create index idx_identifier_asset_subtype on identifier (asset_subtype);" appears 4 times in this file; the first occurrence is on line 29](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import com.dotmarketing.common.db.DotConnect;
+import com.dotmarketing.exception.DotDataException;
+import java.sql.SQLException;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class Task201013AddNewColumnsToIdentifierTableTest {
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java:16]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        final LocalDateTime expectedDate = new Timestamp(
+                expectedResult.getCreateDate().getTime()).toLocalDateTime().truncatedTo(
+                ChronoUnit.SECONDS);
+        final LocalDateTime resultDate = new Timestamp(
+                ((Date)result.get("create_date")).getTime()).toLocalDateTime().truncatedTo(
+                ChronoUnit.SECONDS);
+
+        return result.get("owner").equals(expectedResult.getOwner()) &&
+                resultDate.equals(expectedDate) &&
+                ((!expectedResult.getAssetType().equals("contentlet") && !UtilMethods.isSet(result.get("asset_subtype")))
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:101]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    protected static final String UPDATE_CONTENTLETS =
+            "update identifier set owner=q.i_owner, create_date=q.create_date, asset_subtype=q.asset_subtype from"
+            + "(select cont.identifier, cont.mod_user i_owner, inode.idate create_date, "
+            + "struc.velocity_var_name asset_subtype from contentlet cont, "
+            + "inode inode, structure struc "
+            + "where inode.inode=cont.inode "
+            + "and cont.structure_inode = struc.inode) q where id=q.identifier";
+
+    private static final String UPDATE_FOLDER = "update identifier set owner=q.i_owner, create_date=q.create_date from "
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java:21]: Codacy found an issue: [The String literal "update identifier set owner=q.i_owner, create_date=q.create_date from " appears 4 times in this file; the first occurrence is on line 21](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                        || (expectedResult.getAssetType().equals("contentlet") && result.get("asset_subtype").equals(expectedResult.getAssetSubType())));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:102]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java [hunk 9]
```diff
@@ -317,6 +320,9 @@ protected Identifier createNewIdentifier(final Versionable versionable, final Fo
 				identifier.setURI(uri);
 				identifier.setAssetType(versionable.getVersionType());
 			}
+
+            identifier.setOwner((versionable instanceof WebAsset)
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:324]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+
+public class Task201014UpdateColumnsValuesInIdentifierTableTest {
+
+    private static IdentifierAPI identifierAPI;
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:39]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:101]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java:21]: Codacy found an issue: [The String literal "update identifier set owner=q.i_owner, create_date=q.create_date from " appears 4 times in this file; the first occurrence is on line 21](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:39]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java:16]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:29]: Codacy found an issue: [The String literal "create index idx_identifier_asset_subtype on identifier (asset_subtype);" appears 4 times in this file; the first occurrence is on line 29](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:102]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:324]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:29]: Codacy found an issue: [The String literal "create index idx_identifier_asset_subtype on identifier (asset_subtype);" appears 4 times in this file; the first occurrence is on line 29](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:102]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:324]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java:21]: Codacy found an issue: [The String literal "update identifier set owner=q.i_owner, create_date=q.create_date from " appears 4 times in this file; the first occurrence is on line 21](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:39]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java:16]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:101]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:102]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:101]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:324]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:29]: Codacy found an issue: [The String literal "create index idx_identifier_asset_subtype on identifier (asset_subtype);" appears 4 times in this file; the first occurrence is on line 29](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java:16]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java:21]: Codacy found an issue: [The String literal "update identifier set owner=q.i_owner, create_date=q.create_date from " appears 4 times in this file; the first occurrence is on line 21](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:39]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java [hunk 11]
```diff
@@ -378,19 +398,36 @@ protected Identifier createNewIdentifier ( Versionable versionable, Host site, S
...
 			} else {
                 identifier.setURI( uri );
             }
+
+            identifier.setOwner((versionable instanceof WebAsset)
+                    ? ((WebAsset) versionable).getOwner() : versionable.getModUser());
         }
         identifier.setHostId( site != null ? site.getIdentifier() : null );
 
+        final Inode inode;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:419]: this looks like a candidate to create a private method

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java (hunk 5 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTableTest.java:16]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:102]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 10 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:39]: Codacy found an issue: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java (hunk 7 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTable.java:21]: Codacy found an issue: [The String literal "update identifier set owner=q.i_owner, create_date=q.create_date from " appears 4 times in this file; the first occurrence is on line 21](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java (hunk 4 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task201013AddNewColumnsToIdentifierTable.java:29]: Codacy found an issue: [The String literal "create index idx_identifier_asset_subtype on identifier (asset_subtype);" appears 4 times in this file; the first occurrence is on line 29](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java (hunk 6 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/startup/runonce/Task201014UpdateColumnsValuesInIdentifierTableTest.java:101]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

File: dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/IdentifierFactoryImpl.java:324]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6327767)

**Pull request merged**